Dynamic exclusivity checking bug with nested functions.

**Describe the bug**
A case that probably shouldn't require any dynamic exclusivity checks now has them and fails them incorrectly.

**Steps To Reproduce**
1. Check out the [swift-bug-61404](https://github.com/val-lang/val/tree/swift-bug-61404) tag of the [val repository](https://github.com/val-lang/val).
2. `swift test` for a non-crashing test run with test failures.
3. now revert the last commit (i.e. checkout val-lang/val/58dbd94), which explicitly threads an inout argument through nested functions.
4. `swift test` again for this crash:

```
Test Case '-[ParseGenTests.ParseGenTests testBNFConversion]' started.
Simultaneous accesses to 0x600002adf370, but modification requires exclusive access.
Previous access (a modification) started at
ValPackageTests`EBNF.Grammar.init(_:start:) + 1520 (0x1033799ec).
Current access (a modification) started at:
0    libswiftCore.dylib                 0x00000001a8b6889c swift::runtime::AccessSet::insert(swift::runtime::Access*, void*, void*, swift::ExclusivityFlags) + 432
1    libswiftCore.dylib                 0x00000001a8b68ac4 swift_beginAccess + 84
2    ValPackageTests                    0x0000000103362d20 checkDefined #1 (_:) in EBNF.Grammar.checkAllSymbolsDefined(into:) + 856
3    ValPackageTests                    0x00000001033630dc checkDefined #3 (_:) in EBNF.Grammar.checkAllSymbolsDefined(into:) + 204
4    ValPackageTests                    0x0000000103362a88 checkDefined #2 (_:) in EBNF.Grammar.checkAllSymbolsDefined(into:) + 568
5    ValPackageTests                    0x00000001033627a0 EBNF.Grammar.checkAllSymbolsDefined(into:) + 528
6    ValPackageTests                    0x00000001033793fc EBNF.Grammar.init(_:start:) + 1528
7    ValPackageTests                    0x0000000103389974 String.asEBNFGrammar() + 2860
8    ValPackageTests                    0x000000010338bb58 ParseGenTests.testBNFConversion() + 304
9    ValPackageTests                    0x000000010338c044 @objc ParseGenTests.testBNFConversion() + 36
10   CoreFoundation                     0x000000019b719c70 __invoking___ + 148
11   CoreFoundation                     0x000000019b7199a8 -[NSInvocation invoke] + 468
12   XCTestCore                         0x00000001009de474 +[XCTFailableInvocation invokeStandardConventionInvocation:completion:] + 68
13   XCTestCore                         0x00000001009de450 __90+[XCTFailableInvocation invokeInvocation:withTestMethodConvention:lastObservedErrorIssue:]_block_invoke_3 + 28
14   XCTestCore                         0x00000001009ddd88 __81+[XCTFailableInvocation invokeWithAsynchronousWait:lastObservedErrorIssue:block:]_block_invoke.5 + 112
...
```
**Expected behavior**
`swift test` works with no traps.

**Environment (please fill out the following information)**
 - OS: macOS 12.6
 - Xcode Version/Tag/Branch: Version 14.0.1 (14A400)

**Additional context**
I was unable to reproduce in the small.

@dabrahams the test in question is looking for a file that does not exist in the repo:

Test Case '-[ParseGenTests.ParseGenTests testBNFConversion]' started.
/d/val/Tests/ParseGenTests/ParseGenTests.swift:97: error: -[ParseGenTests.ParseGenTests testBNFConversion] : failed - Unexpected error: Error Domain=NSCocoaErrorDomain Code=260 "The file “spec.md” couldn’t be opened because there is no such file." UserInfo={NSFilePath=/d/val/spec/spec.md, NSUnderlyingError=0x6000030ceb50 {Error Domain=NSPOSIXErrorDomain Code=2 "No such file or directory"}}

--- a/lib/SILOptimizer/Mandatory/AccessEnforcementSelection.cpp
+++ b/lib/SILOptimizer/Mandatory/AccessEnforcementSelection.cpp
@@ -99,6 +99,15 @@ raw_ostream &operator<<(raw_ostream &os, const AddressCapture &capture) {
 
 // For each non-escaping closure, record the indices of arguments that
 // require dynamic enforcement.
+//
+// A note on closure cycles: local functions can be recursive, creating closure
+// cycles. DynamicCaptures ignores such cycles, simply processing the call graph
+// top-down. This relies on a simple rule: if a captured variable is passed as a
+// box a local function (presumably because the function escapes), then it must
+// also be passed as a box to any other local function called by the
+// first. Therefore, if any capture escapes in a closure cycle, then it must be
+// passed as a box in all closures within the cycle. DynamicCaptures does not
+// care about boxes, because they are always dynamically enforced.
 class DynamicCaptures {
   const ClosureFunctionOrder &closureOrder;
 
@@ -127,10 +136,10 @@ public:
   }
 
   bool isDynamic(SILFunctionArgument *arg) const {
-    // If the current function is a local function that directly or indirectly
-    // refers to itself, then conservatively assume dynamic enforcement.
-    if (closureOrder.isHeadOfClosureCycle(arg->getFunction()))
-      return true;
+    // This closure may be the head of a closure cycle. That's ok, because we
+    // only care about whether this argument escapes in the calling function
+    // this is *not* part of the cycle. If the capture escapes anywhere in the
+    // cycle, then it is passed as a box to all closures in that cycle.
 
     auto pos = dynamicCaptureMap.find(arg->getFunction());
     if (pos == dynamicCaptureMap.end())

--- a/test/Interpreter/enforce_exclusive_access.swift
+++ b/test/Interpreter/enforce_exclusive_access.swift
@@ -383,4 +383,56 @@ ExclusiveAccessTestSuite.test("directlyAppliedEscapingConflict") {
   }()
 }
 
+// rdar://102056143 - recursive locals inside a mutating method
+// effectively capture self as inout even if they don't modify
+// self. This should not trigger a runtime trap.
+struct RecursiveLocalDuringMutation {
+  var flag: Bool = false
+
+  mutating func update() {
+    func local1(_ done: Bool) -> Bool {
+      if !done {
+        return local2()
+      }
+      return flag
+    }
+    func local2() -> Bool {
+      return local1(true)
+    }
+    flag = !local1(false)
+  }
+}
+
+ExclusiveAccessTestSuite.test("recursiveLocalDuringMutation") {
+  var v = RecursiveLocalDuringMutation()
+  v.update()
+  _blackHole(v.flag)
+}
+
+// rdar://102056143 - negative test. Make sure we still catch true
+// violations on recursive captures.
+struct RecursiveCaptureViolation {
+  func testBadness() -> Bool {
+    var flag = false
+    func local1(_ x: inout Bool) {
+      if !flag {
+        x = true
+        local2()
+      }
+    }
+    func local2() {
+      local1(&flag)
+    }
+    local2()
+    return flag
+  }
+}
+
+ExclusiveAccessTestSuite.test("recursiveCaptureViolation") {
+  var s = RecursiveCaptureViolation()
+  expectCrashLater()
+  s.testBadness()
+  _blackHole(s)
+}
+
 runAllTests()

--- a/test/SILOptimizer/access_enforcement_selection.swift
+++ b/test/SILOptimizer/access_enforcement_selection.swift
@@ -101,9 +101,8 @@ public func recaptureStack() -> Int {
 
 
 // -----------------------------------------------------------------------------
-// Inout access in a recursive closure currently has dynamic enforcement
-// because enforcement selection cannot visit all closure scopes before
-// selecting enforcement within the closure.
+// Inout access in a recursive non-escaping closure currently has
+// static enforcement.
 
 public protocol Apply {
     func apply(_ body: (Apply) -> ())
@@ -111,7 +110,7 @@ public protocol Apply {
 
 // CHECK-LABEL: sil private @$s28access_enforcement_selection20testRecursiveClosure1a1xyAA5Apply_p_SiztF9localFuncL_1byAaE_p_tF : $@convention(thin) (@in_guaranteed any Apply, @inout_aliasable Int) -> () {
 // CHECK: bb0(%0 : $*any Apply, %1 : @closureCapture $*Int):
-// CHECK:   begin_access [modify] [dynamic] %1 : $*Int
+// CHECK:   begin_access [modify] [static] %1 : $*Int
 // CHECK-LABEL: } // end sil function '$s28access_enforcement_selection20testRecursiveClosure1a1xyAA5Apply_p_SiztF9localFuncL_1byAaE_p_tF'
 public func testRecursiveClosure(a: Apply, x: inout Int) {
     func localFunc(b: Apply) {
@@ -122,4 +121,20 @@ public func testRecursiveClosure(a: Apply, x: inout Int) {
         b.apply(closure)
     }
     a.apply(localFunc)
+}
+
+// CHECK-LABEL: sil private @$s28access_enforcement_selection25testRecursiveLocalCapture1iys5Int64Vz_tF6local2L_yyF : $@convention(thin) (@inout_aliasable Int64) -> () {
+// CHECK: begin_access [modify] [static] %0 : $*Int64
+// CHECK-LABEL: } // end sil function '$s28access_enforcement_selection25testRecursiveLocalCapture1iys5Int64Vz_tF6local2L_yyF'
+func testRecursiveLocalCapture(i: inout Int64) {
+  func local1() {
+    if i < 1 {
+      local2()
+    }
+  }
+  func local2() {
+    i = i + 1
+    local1()
+  }
+  local1()
 }

